### PR TITLE
FL resources

### PR DESCRIPTION
**Pull request by thomasp-ms**
First set of files to support Federated Learning. 

- `automated_provisioning`: resources for automatically provisioning the FL setup
- `fl_arc_k8s`: example using the FL API
- `plan` directory: a generic plan for how to onboard

See more details in the ReadMe's

### CODE REVIEW

File: accelerators/federated-learning/README.md
Comment (dkmiller) [inline accelerators/federated-learning/README.md:3]: Suggestion: add clarification that there isn't yet a committed timeline.

File: accelerators/federated-learning/automated_provisioning/ReadMe.md
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ReadMe.md:43]: Suggestion: remove the default, since this subscription ID won't mean anything outside our team.

File: accelerators/federated-learning/automated_provisioning/ReadMe.md
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ReadMe.md:64]: Same "nit" about removing default value, ditto for e.g. workspace name, resource group, etc.

File: accelerators/federated-learning/automated_provisioning/ReadMe.md
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ReadMe.md:73]: Nit: `ps1` &mapsto; `ps` or `powershell` here and in the other places you type sample PowerShell commands, i.e.

```markdown
    ```powershell
    ./ps/ConnectSiloToOrchestrator.ps1  "..."
    ```
```

File: accelerators/federated-learning/automated_provisioning/ReadMe.md
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ReadMe.md:102]: Suggestion: future work to validate all silos at once.

File: accelerators/federated-learning/automated_provisioning/bicep/AMLWorkspace.bicep
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/bicep/AMLWorkspace.bicep:7]: Suggestion: add some newlines to separate parameters, variables, and resources.

E.g.,

```bicep
@description('Specifies whether to reduce telemetry collection and enable additional encryption.')
param hbi_workspace bool = false

var tenantId = subscription().tenantId

resource storageAccountName 'Microsoft.Storage/storageAccounts@2021-06-01' = {
    ...
```

File: accelerators/federated-learning/automated_provisioning/ReadMe.md
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ReadMe.md:22]: Nit: also need PowerShell, ideally PowerShell Core since that's cross-platform.

File: accelerators/federated-learning/automated_provisioning/ps/Utils.ps1
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ps/Utils.ps1:1]: Nit: PowerShell method names should look like `Verb-OtherStuff`, e.g. below. Ditto for the other method names in this file.

```powershell
Wait-SuccessfulRegistration
```

These articles provide helpful context:

- [Approved Verbs for PowerShell Commands](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.2)
- [Strongly Encouraged Development Guidelines](https://docs.microsoft.com/en-us/powershell/scripting/developer/cmdlet/strongly-encouraged-development-guidelines?view=powershell-7.2)

File: accelerators/federated-learning/automated_provisioning/ps/Utils.ps1
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ps/Utils.ps1:38]: Suggestion: `Write-Error` here since it's an error message.

File: accelerators/federated-learning/automated_provisioning/ps/Utils.ps1
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/ps/Utils.ps1:1]: Nit: `utils` and `helpers` are not great things to name files. Something like `AzureUtilities` is a bit better.

File: accelerators/federated-learning/automated_provisioning/sample_job/RunSampleJob.ps1
Comment (dkmiller) [inline accelerators/federated-learning/automated_provisioning/sample_job/RunSampleJob.ps1:26]: Is there a way you can create this dataset from a remote path? E.g. like

https://github.com/dkmiller/tidbits/blob/master/2021/2021-11-04_mnist-e2e/data/mnist_train_images_raw/data.yml

File: accelerators/federated-learning/fl_arc_k8s/README.md
Comment (dkmiller) [inline accelerators/federated-learning/fl_arc_k8s/README.md:5]: Suggestion: create a Conda environment file capturing these steps. Then the guidelines are as simple as `conda env create --file environment.yml`.

```yaml
name: fl-shrike-examples-env-py38

dependencies:

- python=3.8
- pip
- pip:
  - -r requirements.txt
```

File: accelerators/federated-learning/fl_arc_k8s/components/add_one_thousand_to_parameter/component_spec.yaml
Comment (dkmiller) [inline accelerators/federated-learning/fl_arc_k8s/components/add_one_thousand_to_parameter/component_spec.yaml:8]: Nit: suggest removing tags mentioning `aml-ds@microsoft.com`.

**Review by dkmiller — CHANGES_REQUESTED** [review]

**Review by lostmygithubaccount — CHANGES_REQUESTED** [review]
tagging @balapv 

The extension used in these files is `.yaml`, whereas the rest of the repo (with possible exceptions that slipped in) are using `.yml`. We should be consistent, though the choice to use `.yml` was arbitrarily made by me.

**Comment by dkmiller** [comment]
> tagging @balapv
> 
> The extension used in these files is `.yaml`, whereas the rest of the repo (with possible exceptions that slipped in) are using `.yml`. We should be consistent, though the choice to use `.yml` was arbitrarily made by me.

@lostmygithubaccount , consensus is that `YAML` is better ( https://stackoverflow.com/questions/22268952 ) and some tooling (e.g., [Hydra](http://hydra.cc/)) requires it.

Note, the YAML files aren't DPv2 YAML files, they're config files for libraries being used in defining pipelines.

**Comment by lostmygithubaccount** [comment]
@dkmiller quite a sparse FAQ! https://yaml.org/faq.html 

Understood these aren't DPv2, but we should be consistent across the repository regardless. I'm not personally opposed to switching to `.yaml` at all, but it would require a bit of updating (the DPv2 YAMLs, the GHA workflow definitions, and in certain `readme.py` files that assume `.yml`, possibly other locations). @balapv and @paulshealy1 can help drive the change if agreed on switching.

**Comment by thomasp-ms** [comment]
> @dkmiller quite a sparse FAQ! https://yaml.org/faq.html
> 
> Understood these aren't DPv2, but we should be consistent across the repository regardless. I'm not personally opposed to switching to `.yaml` at all, but it would require a bit of updating (the DPv2 YAMLs, the GHA workflow definitions, and in certain `readme.py` files that assume `.yml`, possibly other locations). @balapv and @paulshealy1 can help drive the change if agreed on switching.

@lostmygithubaccount so you're saying this PR should be blocked until all YAML files in the repo are renamed? It seems like it might not be such a small change. Any way this other can happen after this PR gets merged?

**Comment by lostmygithubaccount** [comment]
@thomasp-ms not saying we should block the PR on this issue, just pointing it out since we tend to let inconsistencies like these into our examples and then never fix/they get worse over time. 

As long as we're aware and plan to fix, not a huge concern -- and not a huge concern regardless for this PR (just want to ensure Bala and Paul are aware of the discussion and potential inconsistency). 

That said, I would like us to have a discussion on where we host accelerators (in this repo or elsewhere) post-holidays with the other Microsoft teams driving that effort. If there's some urgent need to merge this into the main branch, I think it's fine to proceed. We can always move and redirect the content later if we do decide it should live elsewhere.

**Comment by dkmiller** [comment]
Per internal discussion, this content will be added to the following repo instead:

> https://github.com/Azure-Samples/azure-ml-federated-learning/

**Pull request closed without merging**
